plugins/opds.koplugin/opdspse.lua

```diff
@@ -0,0 +1,208 @@
+local http = require("socket.http")
+local ImageViewer = require("ui/widget/imageviewer")
+local InfoMessage = require("ui/widget/infomessage")
+local InputDialog = require("ui/widget/inputdialog")
+local logger = require("logger")
+local ltn12 = require("ltn12")
+local RenderImage = require("ui/renderimage")
+local Screen = require("device").screen
+local socket = require("socket")
+local socketutil = require("socketutil")
+local UIManager = require("ui/uimanager")
+local url = require("socket.url")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+
+
+
+local OPDSPSE = {}
+
+-- This function attempts to pull chapter progress from Kavita.
+function OPDSPSE:getLastPage(item, remote_url)
+    local last_page = 0
+
+    -- create URL's and reference vars
+    local chapter = string.match(remote_url, "chapterId=(%w+)")
+    local api_key = string.match(remote_url, "opds/(.+)/image")
+    local progress_url = string.match(remote_url, "(.+)/api").."/api/Reader/get-progress?chapterId="..chapter
+    local auth_url = string.match(remote_url, "(.+)/api").."/api/Plugin/authenticate?apiKey="..api_key.."&pluginName=KOReader-OPDS"
+
+    -- Do an HTTP POST to get the Bearer Token for authentication of the /api/Reader/get-progress endpoint
+    local auth_parsed = url.parse(auth_url)
+    local auth_data = {}
+    local auth_code, auth_headers, auth_status
+    if auth_parsed.scheme == "http" or auth_parsed.scheme == "https" then
+        socketutil:set_timeout(socketutil.FILE_BLOCK_TIMEOUT, socketutil.FILE_TOTAL_TIMEOUT)
+        auth_code, auth_headers, auth_status = socket.skip(1, http.request {
+            method = "POST",
+            url         = auth_url,
+            headers     = {
+                ["Accept-Encoding"] = "identity",
+                ["Authentication"] = api_key,
+            },
+            sink        = ltn12.sink.table(auth_data),
+            user        = item.username,
+            password    = item.password,

```

Comment:
Does item have username and password fields?
They are not passed down to all items but stored in self, obtained from the root list of servers.
https://github.com/koreader/koreader/blob/c5700b718d8adeadade517ab034c4eee48193f53/plugins/opds.koplugin/opdsbrowser.lua#L913-L917

Reply:
It was originally item.username, but I changed it to root_catalog_username after seeing that change made in the refactor, I think. Should I set it back to item.username and item.password?

Reply:
Yes, if you look in the function in opdsbrowser.lua, `genItemTableFromRoot` adds the catalog username and password to every item.
```
function OPDSBrowser:genItemTableFromRoot()
    local item_table = {}
    -- Loop through the default servers and add them
    -- to the item table.
    for _, server in ipairs(self.opds_servers) do
        table.insert(item_table, {
            text = server.title,
            content = server.subtitle,
            url = server.url,
            username = server.username,
            password = server.password,
            deletable = true,
            editable = true,
            searchable = server.searchable,
        })
    end
```

Reply:
Oh, sorry for leaving more comments, _but_ I have tried this code using OPDS on the demo komga server at demo.komga.org. Still works!

And for future reference, if you want to use the demo komga server for page stream testing, here are the current credentials:
OPDS URL: https://demo.komga.org/opds/v1.2/catalog
Username: demo@komga.org
Password: komga-demo

If for any reason the credentials change, here is a link to that page: https://komga.org/guides/#demo

Reply:
> I'm Sorry, I meant `genItemTableFromCatalog`.
> 
> ```
>  if username then
>         item_table.username = username
>     end
>     if password then
>         item_table.password = password
>     end
> ```

I do not see these lines in the module. Where are they from?

Reply:
Found them in the old version.
No, currently username/password are not stored in items because they are the same for all items in the certain catalog.

Reply:
Alright, I'll try to fix it.

Reply:
> And for future reference, if you want to use the demo komga server for page stream testing, here are the current credentials:

Thanks, tested, works good.